cpp/include/cudf/lists/detail/gather.cuh

```diff
@@ -0,0 +1,290 @@
+ *                          this buffer is not empty, the function will attempt to reuse the
+ *                          buffer if it is large enough.
+ */
+template <bool NullifyOutOfBounds, typename MapItType>
+gather_data make_gather_data(cudf::lists_column_view const& source_column,
+                             MapItType gather_map,
+                             size_type gather_map_size,
+                             cudaStream_t stream,
+                             rmm::mr::device_memory_resource* mr,
+                             rmm::device_uvector<int32_t> prev_base_offsets)
+{
+  // size of the gather map is the # of output rows
+  size_type output_count = gather_map_size;
+  size_type offset_count = output_count + 1;
+
+  // offsets of the source column
+  int32_t const* src_offsets{source_column.offsets().data<int32_t>()};
+  size_type const src_size = source_column.size();
+
+  // outgoing offsets.  these will persist as output from the entire gather operation
+  auto dst_offsets_c = cudf::make_fixed_width_column(
+    data_type{type_id::INT32}, offset_count, mask_state::UNALLOCATED, stream, mr);
+  mutable_column_view dst_offsets_v = dst_offsets_c->mutable_view();
+  int32_t* dst_offsets              = dst_offsets_v.data<int32_t>();
+
+  // generate the compacted outgoing offsets.
+  auto count_iter = thrust::make_counting_iterator<int32_t>(0);
+  thrust::transform_exclusive_scan(
+    rmm::exec_policy(stream)->on(stream),
+    count_iter,
+    count_iter + offset_count,
+    dst_offsets,
+    [gather_map, output_count, src_offsets, src_size] __device__(int32_t index) -> int32_t {
+      int32_t offset_index = index < output_count ? gather_map[index] : 0;
+
+      // if this is an invalid index, this will be a NULL list
+      if (NullifyOutOfBounds && ((offset_index < 0) || (offset_index >= src_size))) { return 0; }
+
+      // the length of this list
+      return src_offsets[offset_index + 1] - src_offsets[offset_index];
+    },
+    0,
+    thrust::plus<int32_t>());
+
+  // For each span of offsets (each output row) we need to know the original offset value to build
+  // the gather map for the next level.  This data is temporary and will only be needed until the
+  // next level of recursion is done using it.
+  //
+  // Additional optimization. If the "prev_base_offsets" buffer we have been passed is large enough
+  // to be used again, re-use it instead of allocating new memory. If it isn't big enough, release
+  // it.
+  if (prev_base_offsets.capacity() > 0) {
+    // reset size (but leave capacity)
+    prev_base_offsets.resize(0, stream);
+  }
+  bool can_recycle = prev_base_offsets.capacity() >= static_cast<size_t>(output_count);
+  rmm::device_uvector<int32_t> base_offsets =
+    can_recycle ? std::move(prev_base_offsets) : rmm::device_uvector<int32_t>(output_count, stream);
+  if (!can_recycle) { prev_base_offsets.release(); }
```

Comment:
I didn't do any formal benchmarking but I did spend some time constructing some test cases.  In practical terms the recycling hardly ever hits : it is very rare that the # of offsets goes -down- as you get deeper into a hierarchy of lists.  

The release of the buffer from the previous level is crucial for keeping total memory usage down (otherwise these potentially huge buffers would remain resident for the duration of the entire recursion stack), but the recycling part is probably not a big win.